{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# importing necessary libraries."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import warnings\n",
    "import seaborn as sns\n",
    "from sklearn import tree\n",
    "from sklearn.feature_selection import SelectKBest\n",
    "from sklearn.feature_selection import chi2\n",
    "from sklearn import preprocessing\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.model_selection import StratifiedKFold\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.metrics import accuracy_score\n",
    "from sklearn.tree import DecisionTreeClassifier\n",
    "from sklearn.naive_bayes import GaussianNB\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# obtaining the data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AF_ESP</th>\n",
       "      <th>AF_EXAC</th>\n",
       "      <th>AF_TGP</th>\n",
       "      <th>CLNDISDB</th>\n",
       "      <th>CLNDISDBINCL</th>\n",
       "      <th>CLNDN</th>\n",
       "      <th>...</th>\n",
       "      <th>SIFT</th>\n",
       "      <th>PolyPhen</th>\n",
       "      <th>MOTIF_NAME</th>\n",
       "      <th>MOTIF_POS</th>\n",
       "      <th>HIGH_INF_POS</th>\n",
       "      <th>MOTIF_SCORE_CHANGE</th>\n",
       "      <th>LoFtool</th>\n",
       "      <th>CADD_PHRED</th>\n",
       "      <th>CADD_RAW</th>\n",
       "      <th>BLOSUM62</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1168180</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0771</td>\n",
       "      <td>0.10020</td>\n",
       "      <td>0.1066</td>\n",
       "      <td>MedGen:CN169374</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not_specified</td>\n",
       "      <td>...</td>\n",
       "      <td>tolerated</td>\n",
       "      <td>benign</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.053</td>\n",
       "      <td>-0.208682</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1470752</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>MedGen:C1843891,OMIM:607454,Orphanet:ORPHA9877...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Spinocerebellar_ataxia_21|not_provided</td>\n",
       "      <td>...</td>\n",
       "      <td>deleterious_low_confidence</td>\n",
       "      <td>benign</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>31.000</td>\n",
       "      <td>6.517838</td>\n",
       "      <td>-3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1737942</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00001</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>Human_Phenotype_Ontology:HP:0000486,MedGen:C00...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Strabismus|Nystagmus|Hypothyroidism|Intellectu...</td>\n",
       "      <td>...</td>\n",
       "      <td>deleterious</td>\n",
       "      <td>probably_damaging</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>28.100</td>\n",
       "      <td>6.061752</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>2160305</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>MedGen:C1321551,OMIM:182212,SNOMED_CT:83092002...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Shprintzen-Goldberg_syndrome|not_provided</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22.500</td>\n",
       "      <td>3.114491</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>2160305</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>MedGen:C1321551,OMIM:182212,SNOMED_CT:83092002</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Shprintzen-Goldberg_syndrome</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>24.700</td>\n",
       "      <td>4.766224</td>\n",
       "      <td>-3.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 46 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  CHROM      POS REF ALT  AF_ESP  AF_EXAC  AF_TGP  \\\n",
       "0     1  1168180   G   C  0.0771  0.10020  0.1066   \n",
       "1     1  1470752   G   A  0.0000  0.00000  0.0000   \n",
       "2     1  1737942   A   G  0.0000  0.00001  0.0000   \n",
       "3     1  2160305   G   A  0.0000  0.00000  0.0000   \n",
       "4     1  2160305   G   T  0.0000  0.00000  0.0000   \n",
       "\n",
       "                                            CLNDISDB CLNDISDBINCL  \\\n",
       "0                                    MedGen:CN169374          NaN   \n",
       "1  MedGen:C1843891,OMIM:607454,Orphanet:ORPHA9877...          NaN   \n",
       "2  Human_Phenotype_Ontology:HP:0000486,MedGen:C00...          NaN   \n",
       "3  MedGen:C1321551,OMIM:182212,SNOMED_CT:83092002...          NaN   \n",
       "4     MedGen:C1321551,OMIM:182212,SNOMED_CT:83092002          NaN   \n",
       "\n",
       "                                               CLNDN  ...  \\\n",
       "0                                      not_specified  ...   \n",
       "1             Spinocerebellar_ataxia_21|not_provided  ...   \n",
       "2  Strabismus|Nystagmus|Hypothyroidism|Intellectu...  ...   \n",
       "3          Shprintzen-Goldberg_syndrome|not_provided  ...   \n",
       "4                       Shprintzen-Goldberg_syndrome  ...   \n",
       "\n",
       "                         SIFT           PolyPhen MOTIF_NAME MOTIF_POS  \\\n",
       "0                   tolerated             benign        NaN       NaN   \n",
       "1  deleterious_low_confidence             benign        NaN       NaN   \n",
       "2                 deleterious  probably_damaging        NaN       NaN   \n",
       "3                         NaN                NaN        NaN       NaN   \n",
       "4                         NaN                NaN        NaN       NaN   \n",
       "\n",
       "  HIGH_INF_POS MOTIF_SCORE_CHANGE  LoFtool  CADD_PHRED  CADD_RAW BLOSUM62  \n",
       "0          NaN                NaN      NaN       1.053 -0.208682      2.0  \n",
       "1          NaN                NaN      NaN      31.000  6.517838     -3.0  \n",
       "2          NaN                NaN      NaN      28.100  6.061752     -1.0  \n",
       "3          NaN                NaN      NaN      22.500  3.114491      NaN  \n",
       "4          NaN                NaN      NaN      24.700  4.766224     -3.0  \n",
       "\n",
       "[5 rows x 46 columns]"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "warnings.filterwarnings('ignore')\n",
    "df=pd.read_csv('clinvar_conflicting.csv')\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# data exploration."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 65188 entries, 0 to 65187\n",
      "Data columns (total 46 columns):\n",
      " #   Column              Non-Null Count  Dtype  \n",
      "---  ------              --------------  -----  \n",
      " 0   CHROM               65188 non-null  object \n",
      " 1   POS                 65188 non-null  int64  \n",
      " 2   REF                 65188 non-null  object \n",
      " 3   ALT                 65188 non-null  object \n",
      " 4   AF_ESP              65188 non-null  float64\n",
      " 5   AF_EXAC             65188 non-null  float64\n",
      " 6   AF_TGP              65188 non-null  float64\n",
      " 7   CLNDISDB            65188 non-null  object \n",
      " 8   CLNDISDBINCL        167 non-null    object \n",
      " 9   CLNDN               65188 non-null  object \n",
      " 10  CLNDNINCL           167 non-null    object \n",
      " 11  CLNHGVS             65188 non-null  object \n",
      " 12  CLNSIGINCL          167 non-null    object \n",
      " 13  CLNVC               65188 non-null  object \n",
      " 14  CLNVI               27659 non-null  object \n",
      " 15  MC                  64342 non-null  object \n",
      " 16  ORIGIN              65188 non-null  int64  \n",
      " 17  SSR                 130 non-null    float64\n",
      " 18  CLASS               65188 non-null  int64  \n",
      " 19  Allele              65188 non-null  object \n",
      " 20  Consequence         65188 non-null  object \n",
      " 21  IMPACT              65188 non-null  object \n",
      " 22  SYMBOL              65172 non-null  object \n",
      " 23  Feature_type        65174 non-null  object \n",
      " 24  Feature             65174 non-null  object \n",
      " 25  BIOTYPE             65172 non-null  object \n",
      " 26  EXON                56295 non-null  object \n",
      " 27  INTRON              8803 non-null   object \n",
      " 28  cDNA_position       56304 non-null  object \n",
      " 29  CDS_position        55233 non-null  object \n",
      " 30  Protein_position    55233 non-null  object \n",
      " 31  Amino_acids         55184 non-null  object \n",
      " 32  Codons              55184 non-null  object \n",
      " 33  DISTANCE            108 non-null    float64\n",
      " 34  STRAND              65174 non-null  float64\n",
      " 35  BAM_EDIT            31969 non-null  object \n",
      " 36  SIFT                24836 non-null  object \n",
      " 37  PolyPhen            24796 non-null  object \n",
      " 38  MOTIF_NAME          2 non-null      object \n",
      " 39  MOTIF_POS           2 non-null      float64\n",
      " 40  HIGH_INF_POS        2 non-null      object \n",
      " 41  MOTIF_SCORE_CHANGE  2 non-null      float64\n",
      " 42  LoFtool             60975 non-null  float64\n",
      " 43  CADD_PHRED          64096 non-null  float64\n",
      " 44  CADD_RAW            64096 non-null  float64\n",
      " 45  BLOSUM62            25593 non-null  float64\n",
      "dtypes: float64(12), int64(3), object(31)\n",
      "memory usage: 22.9+ MB\n"
     ]
    }
   ],
   "source": [
    "df.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>POS</th>\n",
       "      <th>AF_ESP</th>\n",
       "      <th>AF_EXAC</th>\n",
       "      <th>AF_TGP</th>\n",
       "      <th>ORIGIN</th>\n",
       "      <th>SSR</th>\n",
       "      <th>CLASS</th>\n",
       "      <th>DISTANCE</th>\n",
       "      <th>STRAND</th>\n",
       "      <th>MOTIF_POS</th>\n",
       "      <th>MOTIF_SCORE_CHANGE</th>\n",
       "      <th>LoFtool</th>\n",
       "      <th>CADD_PHRED</th>\n",
       "      <th>CADD_RAW</th>\n",
       "      <th>BLOSUM62</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>POS</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.007003</td>\n",
       "      <td>0.007721</td>\n",
       "      <td>0.004225</td>\n",
       "      <td>0.013645</td>\n",
       "      <td>0.192398</td>\n",
       "      <td>0.012061</td>\n",
       "      <td>0.197942</td>\n",
       "      <td>-0.122896</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.257226</td>\n",
       "      <td>-0.005409</td>\n",
       "      <td>-0.009753</td>\n",
       "      <td>0.022451</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AF_ESP</th>\n",
       "      <td>0.007003</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.851870</td>\n",
       "      <td>0.807741</td>\n",
       "      <td>-0.013563</td>\n",
       "      <td>-0.077253</td>\n",
       "      <td>-0.126538</td>\n",
       "      <td>0.295624</td>\n",
       "      <td>-0.011551</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.027343</td>\n",
       "      <td>-0.164169</td>\n",
       "      <td>-0.148118</td>\n",
       "      <td>0.039367</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AF_EXAC</th>\n",
       "      <td>0.007721</td>\n",
       "      <td>0.851870</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.805634</td>\n",
       "      <td>-0.013432</td>\n",
       "      <td>0.048413</td>\n",
       "      <td>-0.121213</td>\n",
       "      <td>0.291840</td>\n",
       "      <td>-0.013358</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.027831</td>\n",
       "      <td>-0.154891</td>\n",
       "      <td>-0.140803</td>\n",
       "      <td>0.038996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AF_TGP</th>\n",
       "      <td>0.004225</td>\n",
       "      <td>0.807741</td>\n",
       "      <td>0.805634</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.014450</td>\n",
       "      <td>-0.114526</td>\n",
       "      <td>-0.130077</td>\n",
       "      <td>0.040676</td>\n",
       "      <td>-0.014002</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.029447</td>\n",
       "      <td>-0.167057</td>\n",
       "      <td>-0.150989</td>\n",
       "      <td>0.043167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORIGIN</th>\n",
       "      <td>0.013645</td>\n",
       "      <td>-0.013563</td>\n",
       "      <td>-0.013432</td>\n",
       "      <td>-0.014450</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.175675</td>\n",
       "      <td>0.004357</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.007564</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.028363</td>\n",
       "      <td>0.060290</td>\n",
       "      <td>0.064609</td>\n",
       "      <td>-0.012681</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              POS    AF_ESP   AF_EXAC    AF_TGP    ORIGIN       SSR     CLASS  \\\n",
       "POS      1.000000  0.007003  0.007721  0.004225  0.013645  0.192398  0.012061   \n",
       "AF_ESP   0.007003  1.000000  0.851870  0.807741 -0.013563 -0.077253 -0.126538   \n",
       "AF_EXAC  0.007721  0.851870  1.000000  0.805634 -0.013432  0.048413 -0.121213   \n",
       "AF_TGP   0.004225  0.807741  0.805634  1.000000 -0.014450 -0.114526 -0.130077   \n",
       "ORIGIN   0.013645 -0.013563 -0.013432 -0.014450  1.000000  0.175675  0.004357   \n",
       "\n",
       "         DISTANCE    STRAND  MOTIF_POS  MOTIF_SCORE_CHANGE   LoFtool  \\\n",
       "POS      0.197942 -0.122896        NaN                -1.0  0.257226   \n",
       "AF_ESP   0.295624 -0.011551        NaN                 NaN  0.027343   \n",
       "AF_EXAC  0.291840 -0.013358        NaN                 NaN  0.027831   \n",
       "AF_TGP   0.040676 -0.014002        NaN                 NaN  0.029447   \n",
       "ORIGIN        NaN -0.007564        NaN                 NaN -0.028363   \n",
       "\n",
       "         CADD_PHRED  CADD_RAW  BLOSUM62  \n",
       "POS       -0.005409 -0.009753  0.022451  \n",
       "AF_ESP    -0.164169 -0.148118  0.039367  \n",
       "AF_EXAC   -0.154891 -0.140803  0.038996  \n",
       "AF_TGP    -0.167057 -0.150989  0.043167  \n",
       "ORIGIN     0.060290  0.064609 -0.012681  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(df.corr(numeric_only=True).head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# dropping the useless columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.drop(['CLNDISDBINCL','CLNDNINCL','CLNSIGINCL','CLNVI','SSR','INTRON','DISTANCE','BAM_EDIT','SIFT','PolyPhen','MOTIF_NAME','MOTIF_POS','HIGH_INF_POS','MOTIF_SCORE_CHANGE','BLOSUM62'],axis=1,inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CHROM                   0\n",
       "POS                     0\n",
       "REF                     0\n",
       "ALT                     0\n",
       "AF_ESP                  0\n",
       "AF_EXAC                 0\n",
       "AF_TGP                  0\n",
       "CLNDISDB                0\n",
       "CLNDN                   0\n",
       "CLNHGVS                 0\n",
       "CLNVC                   0\n",
       "MC                    846\n",
       "ORIGIN                  0\n",
       "CLASS                   0\n",
       "Allele                  0\n",
       "Consequence             0\n",
       "IMPACT                  0\n",
       "SYMBOL                 16\n",
       "Feature_type           14\n",
       "Feature                14\n",
       "BIOTYPE                16\n",
       "EXON                 8893\n",
       "cDNA_position        8884\n",
       "CDS_position         9955\n",
       "Protein_position     9955\n",
       "Amino_acids         10004\n",
       "Codons              10004\n",
       "STRAND                 14\n",
       "LoFtool              4213\n",
       "CADD_PHRED           1092\n",
       "CADD_RAW             1092\n",
       "dtype: int64"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.isnull().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['LoFtool'].fillna(int(df['LoFtool'].mean()), inplace=True)\n",
    "df['CADD_PHRED'].fillna(int(df['CADD_PHRED'].mean()), inplace=True)\n",
    "df['CADD_RAW'].fillna(int(df['CADD_RAW'].mean()), inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['STRAND'].fillna(df['STRAND'].mode()[0],inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['MC'].fillna(df['MC'].mode()[0],inplace=True)\n",
    "df['SYMBOL'].fillna(df['SYMBOL'].mode()[0],inplace=True)\n",
    "df['Feature_type'].fillna(df['Feature_type'].mode()[0],inplace=True)\n",
    "df['Feature'].fillna(df['Feature'].mode()[0],inplace=True)\n",
    "df['BIOTYPE'].fillna(df['BIOTYPE'].mode()[0],inplace=True)\n",
    "df['EXON'].fillna(df['EXON'].mode()[0],inplace=True)\n",
    "df['cDNA_position'].fillna(df['cDNA_position'].mode()[0],inplace=True)\n",
    "df['CDS_position'].fillna(df['CDS_position'].mode()[0],inplace=True)\n",
    "df['Protein_position'].fillna(df['Protein_position'].mode()[0],inplace=True)\n",
    "df['Amino_acids'].fillna(df['Amino_acids'].mode()[0],inplace=True)\n",
    "df['Codons'].fillna(df['Codons'].mode()[0],inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CHROM               0\n",
       "POS                 0\n",
       "REF                 0\n",
       "ALT                 0\n",
       "AF_ESP              0\n",
       "AF_EXAC             0\n",
       "AF_TGP              0\n",
       "CLNDISDB            0\n",
       "CLNDN               0\n",
       "CLNHGVS             0\n",
       "CLNVC               0\n",
       "MC                  0\n",
       "ORIGIN              0\n",
       "CLASS               0\n",
       "Allele              0\n",
       "Consequence         0\n",
       "IMPACT              0\n",
       "SYMBOL              0\n",
       "Feature_type        0\n",
       "Feature             0\n",
       "BIOTYPE             0\n",
       "EXON                0\n",
       "cDNA_position       0\n",
       "CDS_position        0\n",
       "Protein_position    0\n",
       "Amino_acids         0\n",
       "Codons              0\n",
       "STRAND              0\n",
       "LoFtool             0\n",
       "CADD_PHRED          0\n",
       "CADD_RAW            0\n",
       "dtype: int64"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.isnull().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "object_columns = df.select_dtypes(['object']).columns\n",
    "object_columns\n",
    "for c in df.columns:\n",
    "    if c in object_columns:\n",
    "        df[[c]]=df[[c]].astype('category')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['CHROM', 'REF', 'ALT', 'CLNDISDB', 'CLNDN', 'CLNHGVS', 'CLNVC', 'MC',\n",
      "       'Allele', 'Consequence', 'IMPACT', 'SYMBOL', 'Feature_type', 'Feature',\n",
      "       'BIOTYPE', 'EXON', 'cDNA_position', 'CDS_position', 'Protein_position',\n",
      "       'Amino_acids', 'Codons'],\n",
      "      dtype='object')\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Index(['CHROM', 'REF', 'ALT', 'CLNDISDB', 'CLNDN', 'CLNHGVS', 'CLNVC', 'MC',\n",
       "       'Allele', 'Consequence', 'IMPACT', 'SYMBOL', 'Feature_type', 'Feature',\n",
       "       'BIOTYPE', 'EXON', 'cDNA_position', 'CDS_position', 'Protein_position',\n",
       "       'Amino_acids', 'Codons'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cat_columns = df.select_dtypes(['category']).columns\n",
    "print(cat_columns)\n",
    "df[cat_columns] = df[cat_columns].apply(lambda x: x.cat.codes)\n",
    "cat_columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CHROM                  int8\n",
       "POS                   int64\n",
       "REF                   int16\n",
       "ALT                   int16\n",
       "AF_ESP              float64\n",
       "AF_EXAC             float64\n",
       "AF_TGP              float64\n",
       "CLNDISDB              int16\n",
       "CLNDN                 int16\n",
       "CLNHGVS               int32\n",
       "CLNVC                  int8\n",
       "MC                     int8\n",
       "ORIGIN                int64\n",
       "CLASS                 int64\n",
       "Allele                int16\n",
       "Consequence            int8\n",
       "IMPACT                 int8\n",
       "SYMBOL                int16\n",
       "Feature_type           int8\n",
       "Feature               int16\n",
       "BIOTYPE                int8\n",
       "EXON                  int16\n",
       "cDNA_position         int16\n",
       "CDS_position          int16\n",
       "Protein_position      int16\n",
       "Amino_acids           int16\n",
       "Codons                int16\n",
       "STRAND              float64\n",
       "LoFtool             float64\n",
       "CADD_PHRED          float64\n",
       "CADD_RAW            float64\n",
       "dtype: object"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CHROM               0.002030\n",
       "POS                 0.012061\n",
       "REF                 0.000586\n",
       "ALT                -0.000748\n",
       "AF_ESP             -0.126538\n",
       "AF_EXAC            -0.121213\n",
       "AF_TGP             -0.130077\n",
       "CLNDISDB            0.012407\n",
       "CLNDN               0.023176\n",
       "CLNHGVS            -0.013249\n",
       "CLNVC               0.037792\n",
       "MC                  0.006453\n",
       "ORIGIN              0.004357\n",
       "CLASS               1.000000\n",
       "Allele              0.007442\n",
       "Consequence        -0.001063\n",
       "IMPACT              0.068912\n",
       "SYMBOL              0.022337\n",
       "Feature_type       -0.009541\n",
       "Feature             0.001686\n",
       "BIOTYPE             0.001276\n",
       "EXON               -0.010719\n",
       "cDNA_position       0.001162\n",
       "CDS_position       -0.013194\n",
       "Protein_position   -0.007400\n",
       "Amino_acids        -0.016488\n",
       "Codons             -0.004252\n",
       "STRAND             -0.031551\n",
       "LoFtool            -0.002226\n",
       "CADD_PHRED         -0.037178\n",
       "CADD_RAW           -0.052208\n",
       "Name: CLASS, dtype: float64"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.corr()['CLASS']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# feature selection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "original_df=df.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AF_ESP</th>\n",
       "      <th>AF_EXAC</th>\n",
       "      <th>AF_TGP</th>\n",
       "      <th>CLNDISDB</th>\n",
       "      <th>CLNDN</th>\n",
       "      <th>CLNHGVS</th>\n",
       "      <th>...</th>\n",
       "      <th>EXON</th>\n",
       "      <th>cDNA_position</th>\n",
       "      <th>CDS_position</th>\n",
       "      <th>Protein_position</th>\n",
       "      <th>Amino_acids</th>\n",
       "      <th>Codons</th>\n",
       "      <th>STRAND</th>\n",
       "      <th>LoFtool</th>\n",
       "      <th>CADD_PHRED</th>\n",
       "      <th>CADD_RAW</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.378378</td>\n",
       "      <td>0.004714</td>\n",
       "      <td>0.505202</td>\n",
       "      <td>0.231947</td>\n",
       "      <td>0.154509</td>\n",
       "      <td>0.200444</td>\n",
       "      <td>0.213285</td>\n",
       "      <td>0.960251</td>\n",
       "      <td>0.967707</td>\n",
       "      <td>0.005001</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.636839</td>\n",
       "      <td>0.608696</td>\n",
       "      <td>0.198283</td>\n",
       "      <td>0.203807</td>\n",
       "      <td>0.744930</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.010626</td>\n",
       "      <td>0.101256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.378378</td>\n",
       "      <td>0.005936</td>\n",
       "      <td>0.505202</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.716452</td>\n",
       "      <td>0.905929</td>\n",
       "      <td>0.007164</td>\n",
       "      <td>...</td>\n",
       "      <td>0.624272</td>\n",
       "      <td>0.606343</td>\n",
       "      <td>0.595813</td>\n",
       "      <td>0.188198</td>\n",
       "      <td>0.599524</td>\n",
       "      <td>0.515097</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.313124</td>\n",
       "      <td>0.230528</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.378378</td>\n",
       "      <td>0.007015</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.466083</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000020</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.081230</td>\n",
       "      <td>0.912518</td>\n",
       "      <td>0.018086</td>\n",
       "      <td>...</td>\n",
       "      <td>0.814281</td>\n",
       "      <td>0.713222</td>\n",
       "      <td>0.292637</td>\n",
       "      <td>0.928591</td>\n",
       "      <td>0.380650</td>\n",
       "      <td>0.366381</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.283831</td>\n",
       "      <td>0.221763</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.378378</td>\n",
       "      <td>0.008721</td>\n",
       "      <td>0.505202</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.649085</td>\n",
       "      <td>0.880873</td>\n",
       "      <td>0.028012</td>\n",
       "      <td>...</td>\n",
       "      <td>0.016243</td>\n",
       "      <td>0.115685</td>\n",
       "      <td>0.000293</td>\n",
       "      <td>0.624012</td>\n",
       "      <td>0.298969</td>\n",
       "      <td>0.235692</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.227265</td>\n",
       "      <td>0.165122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.378378</td>\n",
       "      <td>0.008721</td>\n",
       "      <td>0.505202</td>\n",
       "      <td>0.733042</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.648652</td>\n",
       "      <td>0.880657</td>\n",
       "      <td>0.028027</td>\n",
       "      <td>...</td>\n",
       "      <td>0.016243</td>\n",
       "      <td>0.115685</td>\n",
       "      <td>0.000293</td>\n",
       "      <td>0.624012</td>\n",
       "      <td>0.287074</td>\n",
       "      <td>0.236593</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.249487</td>\n",
       "      <td>0.196865</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65183</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.622588</td>\n",
       "      <td>0.744509</td>\n",
       "      <td>0.466083</td>\n",
       "      <td>0.160521</td>\n",
       "      <td>0.278521</td>\n",
       "      <td>0.321128</td>\n",
       "      <td>0.346041</td>\n",
       "      <td>0.422832</td>\n",
       "      <td>0.983171</td>\n",
       "      <td>...</td>\n",
       "      <td>0.132087</td>\n",
       "      <td>0.483428</td>\n",
       "      <td>0.467940</td>\n",
       "      <td>0.077133</td>\n",
       "      <td>0.767645</td>\n",
       "      <td>0.938711</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00158</td>\n",
       "      <td>0.001051</td>\n",
       "      <td>0.093141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65184</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.622592</td>\n",
       "      <td>0.258960</td>\n",
       "      <td>0.733042</td>\n",
       "      <td>0.004008</td>\n",
       "      <td>0.001200</td>\n",
       "      <td>0.002601</td>\n",
       "      <td>0.967400</td>\n",
       "      <td>0.981207</td>\n",
       "      <td>0.983218</td>\n",
       "      <td>...</td>\n",
       "      <td>0.132087</td>\n",
       "      <td>0.382490</td>\n",
       "      <td>0.365613</td>\n",
       "      <td>0.993322</td>\n",
       "      <td>0.915147</td>\n",
       "      <td>0.241100</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00158</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.071990</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65185</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.622737</td>\n",
       "      <td>0.258960</td>\n",
       "      <td>0.733042</td>\n",
       "      <td>0.025050</td>\n",
       "      <td>0.007402</td>\n",
       "      <td>0.022209</td>\n",
       "      <td>0.967400</td>\n",
       "      <td>0.981207</td>\n",
       "      <td>0.983264</td>\n",
       "      <td>...</td>\n",
       "      <td>0.935642</td>\n",
       "      <td>0.097287</td>\n",
       "      <td>0.048016</td>\n",
       "      <td>0.671573</td>\n",
       "      <td>0.061856</td>\n",
       "      <td>0.789995</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00158</td>\n",
       "      <td>0.129789</td>\n",
       "      <td>0.132411</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65186</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.623929</td>\n",
       "      <td>0.744509</td>\n",
       "      <td>0.231947</td>\n",
       "      <td>0.000601</td>\n",
       "      <td>0.000680</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.890393</td>\n",
       "      <td>0.721784</td>\n",
       "      <td>0.983279</td>\n",
       "      <td>...</td>\n",
       "      <td>0.254980</td>\n",
       "      <td>0.869998</td>\n",
       "      <td>0.631313</td>\n",
       "      <td>0.219815</td>\n",
       "      <td>0.848533</td>\n",
       "      <td>0.398828</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.001303</td>\n",
       "      <td>0.093881</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65187</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.624003</td>\n",
       "      <td>0.505202</td>\n",
       "      <td>0.231947</td>\n",
       "      <td>0.003808</td>\n",
       "      <td>0.005341</td>\n",
       "      <td>0.001601</td>\n",
       "      <td>0.991660</td>\n",
       "      <td>0.999352</td>\n",
       "      <td>0.983294</td>\n",
       "      <td>...</td>\n",
       "      <td>0.829911</td>\n",
       "      <td>0.843869</td>\n",
       "      <td>0.619821</td>\n",
       "      <td>0.210548</td>\n",
       "      <td>0.594766</td>\n",
       "      <td>0.150068</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.14000</td>\n",
       "      <td>0.000455</td>\n",
       "      <td>0.090151</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>65188 rows × 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          CHROM       POS       REF       ALT    AF_ESP   AF_EXAC    AF_TGP  \\\n",
       "0      0.378378  0.004714  0.505202  0.231947  0.154509  0.200444  0.213285   \n",
       "1      0.378378  0.005936  0.505202  0.000000  0.000000  0.000000  0.000000   \n",
       "2      0.378378  0.007015  0.000000  0.466083  0.000000  0.000020  0.000000   \n",
       "3      0.378378  0.008721  0.505202  0.000000  0.000000  0.000000  0.000000   \n",
       "4      0.378378  0.008721  0.505202  0.733042  0.000000  0.000000  0.000000   \n",
       "...         ...       ...       ...       ...       ...       ...       ...   \n",
       "65183  1.000000  0.622588  0.744509  0.466083  0.160521  0.278521  0.321128   \n",
       "65184  1.000000  0.622592  0.258960  0.733042  0.004008  0.001200  0.002601   \n",
       "65185  1.000000  0.622737  0.258960  0.733042  0.025050  0.007402  0.022209   \n",
       "65186  1.000000  0.623929  0.744509  0.231947  0.000601  0.000680  0.000000   \n",
       "65187  1.000000  0.624003  0.505202  0.231947  0.003808  0.005341  0.001601   \n",
       "\n",
       "       CLNDISDB     CLNDN   CLNHGVS  ...      EXON  cDNA_position  \\\n",
       "0      0.960251  0.967707  0.005001  ...  0.000000       0.636839   \n",
       "1      0.716452  0.905929  0.007164  ...  0.624272       0.606343   \n",
       "2      0.081230  0.912518  0.018086  ...  0.814281       0.713222   \n",
       "3      0.649085  0.880873  0.028012  ...  0.016243       0.115685   \n",
       "4      0.648652  0.880657  0.028027  ...  0.016243       0.115685   \n",
       "...         ...       ...       ...  ...       ...            ...   \n",
       "65183  0.346041  0.422832  0.983171  ...  0.132087       0.483428   \n",
       "65184  0.967400  0.981207  0.983218  ...  0.132087       0.382490   \n",
       "65185  0.967400  0.981207  0.983264  ...  0.935642       0.097287   \n",
       "65186  0.890393  0.721784  0.983279  ...  0.254980       0.869998   \n",
       "65187  0.991660  0.999352  0.983294  ...  0.829911       0.843869   \n",
       "\n",
       "       CDS_position  Protein_position  Amino_acids    Codons  STRAND  LoFtool  \\\n",
       "0          0.608696          0.198283     0.203807  0.744930     1.0  0.00000   \n",
       "1          0.595813          0.188198     0.599524  0.515097     0.0  0.00000   \n",
       "2          0.292637          0.928591     0.380650  0.366381     0.0  0.00000   \n",
       "3          0.000293          0.624012     0.298969  0.235692     1.0  0.00000   \n",
       "4          0.000293          0.624012     0.287074  0.236593     1.0  0.00000   \n",
       "...             ...               ...          ...       ...     ...      ...   \n",
       "65183      0.467940          0.077133     0.767645  0.938711     0.0  0.00158   \n",
       "65184      0.365613          0.993322     0.915147  0.241100     0.0  0.00158   \n",
       "65185      0.048016          0.671573     0.061856  0.789995     0.0  0.00158   \n",
       "65186      0.631313          0.219815     0.848533  0.398828     0.0  0.00000   \n",
       "65187      0.619821          0.210548     0.594766  0.150068     0.0  0.14000   \n",
       "\n",
       "       CADD_PHRED  CADD_RAW  \n",
       "0        0.010626  0.101256  \n",
       "1        0.313124  0.230528  \n",
       "2        0.283831  0.221763  \n",
       "3        0.227265  0.165122  \n",
       "4        0.249487  0.196865  \n",
       "...           ...       ...  \n",
       "65183    0.001051  0.093141  \n",
       "65184    0.000010  0.071990  \n",
       "65185    0.129789  0.132411  \n",
       "65186    0.001303  0.093881  \n",
       "65187    0.000455  0.090151  \n",
       "\n",
       "[65188 rows x 31 columns]"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# performing min max scaling so that we can get values in the range of 0 to 1 as chi square doesn't work on negative values\n",
    "scaler=preprocessing.MinMaxScaler()\n",
    "XScaled=scaler.fit_transform(df)\n",
    "scaled_df=pd.DataFrame(XScaled,columns=df.columns)\n",
    "scaled_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "         Specs       Score\n",
      "6       AF_TGP  512.323040\n",
      "4       AF_ESP  481.512304\n",
      "5      AF_EXAC  468.705551\n",
      "15      IMPACT   35.646297\n",
      "26      STRAND   32.669229\n",
      "28  CADD_PHRED    6.705156\n",
      "16      SYMBOL    6.215260\n",
      "8        CLNDN    5.866560\n"
     ]
    }
   ],
   "source": [
    "# using chi square for feature selection\n",
    "data=scaled_df.copy()\n",
    "X = data.loc[:,data.columns!='CLASS']\n",
    "y = data.loc[:,'CLASS']\n",
    "bestfeatures = SelectKBest(score_func=chi2, k='all')\n",
    "fit = bestfeatures.fit(X,y)\n",
    "dfscores = pd.DataFrame(fit.scores_)\n",
    "dfcolumns = pd.DataFrame(X.columns)\n",
    "featureScores = pd.concat([dfcolumns,dfscores],axis=1)\n",
    "featureScores.columns = ['Specs','Score']\n",
    "print(featureScores.nlargest(8,'Score'))#printing the 8 best features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AF_TGP</th>\n",
       "      <th>AF_ESP</th>\n",
       "      <th>AF_EXAC</th>\n",
       "      <th>IMPACT</th>\n",
       "      <th>STRAND</th>\n",
       "      <th>CADD_PHRED</th>\n",
       "      <th>SYMBOL</th>\n",
       "      <th>CLNDN</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.213285</td>\n",
       "      <td>0.154509</td>\n",
       "      <td>0.200444</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.010626</td>\n",
       "      <td>0.085948</td>\n",
       "      <td>0.967707</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.313124</td>\n",
       "      <td>0.917490</td>\n",
       "      <td>0.905929</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000020</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.283831</td>\n",
       "      <td>0.383326</td>\n",
       "      <td>0.912518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.227265</td>\n",
       "      <td>0.811775</td>\n",
       "      <td>0.880873</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.249487</td>\n",
       "      <td>0.811775</td>\n",
       "      <td>0.880657</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65183</th>\n",
       "      <td>0.321128</td>\n",
       "      <td>0.160521</td>\n",
       "      <td>0.278521</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001051</td>\n",
       "      <td>0.303825</td>\n",
       "      <td>0.422832</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65184</th>\n",
       "      <td>0.002601</td>\n",
       "      <td>0.004008</td>\n",
       "      <td>0.001200</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.303825</td>\n",
       "      <td>0.981207</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65185</th>\n",
       "      <td>0.022209</td>\n",
       "      <td>0.025050</td>\n",
       "      <td>0.007402</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.129789</td>\n",
       "      <td>0.303825</td>\n",
       "      <td>0.981207</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65186</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000601</td>\n",
       "      <td>0.000680</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001303</td>\n",
       "      <td>0.735281</td>\n",
       "      <td>0.721784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65187</th>\n",
       "      <td>0.001601</td>\n",
       "      <td>0.003808</td>\n",
       "      <td>0.005341</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000455</td>\n",
       "      <td>0.177482</td>\n",
       "      <td>0.999352</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>65188 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         AF_TGP    AF_ESP   AF_EXAC    IMPACT  STRAND  CADD_PHRED    SYMBOL  \\\n",
       "0      0.213285  0.154509  0.200444  0.666667     1.0    0.010626  0.085948   \n",
       "1      0.000000  0.000000  0.000000  0.666667     0.0    0.313124  0.917490   \n",
       "2      0.000000  0.000000  0.000020  0.666667     0.0    0.283831  0.383326   \n",
       "3      0.000000  0.000000  0.000000  0.666667     1.0    0.227265  0.811775   \n",
       "4      0.000000  0.000000  0.000000  0.666667     1.0    0.249487  0.811775   \n",
       "...         ...       ...       ...       ...     ...         ...       ...   \n",
       "65183  0.321128  0.160521  0.278521  0.333333     0.0    0.001051  0.303825   \n",
       "65184  0.002601  0.004008  0.001200  0.666667     0.0    0.000010  0.303825   \n",
       "65185  0.022209  0.025050  0.007402  0.333333     0.0    0.129789  0.303825   \n",
       "65186  0.000000  0.000601  0.000680  0.333333     0.0    0.001303  0.735281   \n",
       "65187  0.001601  0.003808  0.005341  0.666667     0.0    0.000455  0.177482   \n",
       "\n",
       "          CLNDN  \n",
       "0      0.967707  \n",
       "1      0.905929  \n",
       "2      0.912518  \n",
       "3      0.880873  \n",
       "4      0.880657  \n",
       "...         ...  \n",
       "65183  0.422832  \n",
       "65184  0.981207  \n",
       "65185  0.981207  \n",
       "65186  0.721784  \n",
       "65187  0.999352  \n",
       "\n",
       "[65188 rows x 8 columns]"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X = data[featureScores.sort_values(by='Score', ascending=False).iloc[:8,0]]\n",
    "y = data['CLASS']\n",
    "X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "accsLogistic = []\n",
    "skf = StratifiedKFold(n_splits=10,random_state=1, shuffle=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# prediciton using Logistic Regression Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Test Accuracy 74.79674796747967\n",
      "Test Accuracy 74.79674796747967\n",
      "Test Accuracy 74.79674796747967\n",
      "Test Accuracy 74.79674796747967\n",
      "Test Accuracy 74.7814081914404\n",
      "Test Accuracy 74.7814081914404\n",
      "Test Accuracy 74.7814081914404\n",
      "Test Accuracy 74.7814081914404\n",
      "Test Accuracy 74.79288125191776\n",
      "Test Accuracy 74.79288125191776\n",
      "accuracies in iterations :  [74.79674796747967, 74.79674796747967, 74.79674796747967, 74.79674796747967, 74.7814081914404, 74.7814081914404, 74.7814081914404, 74.7814081914404, 74.79288125191776, 74.79288125191776]\n",
      "AVG:  74.78983871395158\n"
     ]
    }
   ],
   "source": [
    "for train_index, test_index in skf.split(X,y):\n",
    "    X_train = X.iloc[train_index]\n",
    "    y_train = y[train_index]\n",
    "    X_test = X.iloc[test_index]\n",
    "    y_test = y[test_index]\n",
    "    LRModel = LogisticRegression(penalty = 'l2',max_iter=500)\n",
    "    LRModel.fit(X_train, y_train)\n",
    "    Y_testPred = LRModel.predict(X_test)\n",
    "    testAccuracy = accuracy_score(y_test, Y_testPred)\n",
    "    print(\"Test Accuracy\", testAccuracy*100)\n",
    "    accsLogistic.append(testAccuracy*100)\n",
    "print(\"accuracies in iterations : \",accsLogistic)\n",
    "print(\"AVG: \",sum(accsLogistic)/len(accsLogistic))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# prediction using Decision Tree Classifier"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "accsDecisionTreeClasifier = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Test Accuracy 72.32704402515722\n",
      "Test Accuracy 72.74121797821752\n",
      "Test Accuracy 71.92820984813622\n",
      "Test Accuracy 71.43733701487959\n",
      "Test Accuracy 71.6827734315079\n",
      "Test Accuracy 72.95597484276729\n",
      "Test Accuracy 71.80549163982207\n",
      "Test Accuracy 72.14296671268599\n",
      "Test Accuracy 72.24608775698067\n",
      "Test Accuracy 72.18471923903039\n",
      "accuracies in iterations :  [72.32704402515722, 72.74121797821752, 71.92820984813622, 71.43733701487959, 71.6827734315079, 72.95597484276729, 71.80549163982207, 72.14296671268599, 72.24608775698067, 72.18471923903039]\n",
      "AVG:  72.14518224891847\n"
     ]
    }
   ],
   "source": [
    "# using Decision Tree Classifier\n",
    "dtree = DecisionTreeClassifier()\n",
    "for train_index, test_index in skf.split(X,y):\n",
    "    X_train = X.iloc[train_index]\n",
    "    y_train = y[train_index]\n",
    "    X_test = X.iloc[test_index]\n",
    "    y_test = y[test_index]\n",
    "    dtree.fit(X_train, y_train)\n",
    "    Y_testPred = dtree.predict(X_test)\n",
    "    testAccuracy = accuracy_score(y_test, Y_testPred)\n",
    "    print(\"Test Accuracy\", testAccuracy*100)\n",
    "    accsDecisionTreeClasifier.append(testAccuracy*100)\n",
    "print(\"accuracies in iterations : \",accsDecisionTreeClasifier)\n",
    "print(\"AVG: \",sum(accsDecisionTreeClasifier)/len(accsDecisionTreeClasifier))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# prediction using Naive Bayes Classifier"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [],
   "source": [
    "accsNaiveBayesClassifier = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AF_TGP</th>\n",
       "      <th>AF_ESP</th>\n",
       "      <th>AF_EXAC</th>\n",
       "      <th>IMPACT</th>\n",
       "      <th>STRAND</th>\n",
       "      <th>CADD_PHRED</th>\n",
       "      <th>SYMBOL</th>\n",
       "      <th>CLNDN</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.1066</td>\n",
       "      <td>0.0771</td>\n",
       "      <td>0.10020</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.053</td>\n",
       "      <td>200</td>\n",
       "      <td>8960</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>2</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>31.000</td>\n",
       "      <td>2135</td>\n",
       "      <td>8388</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00001</td>\n",
       "      <td>2</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>28.100</td>\n",
       "      <td>892</td>\n",
       "      <td>8449</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>22.500</td>\n",
       "      <td>1889</td>\n",
       "      <td>8156</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>24.700</td>\n",
       "      <td>1889</td>\n",
       "      <td>8154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65183</th>\n",
       "      <td>0.1605</td>\n",
       "      <td>0.0801</td>\n",
       "      <td>0.13923</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.105</td>\n",
       "      <td>707</td>\n",
       "      <td>3915</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65184</th>\n",
       "      <td>0.0013</td>\n",
       "      <td>0.0020</td>\n",
       "      <td>0.00060</td>\n",
       "      <td>2</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.002</td>\n",
       "      <td>707</td>\n",
       "      <td>9085</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65185</th>\n",
       "      <td>0.0111</td>\n",
       "      <td>0.0125</td>\n",
       "      <td>0.00370</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>12.850</td>\n",
       "      <td>707</td>\n",
       "      <td>9085</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65186</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0003</td>\n",
       "      <td>0.00034</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.130</td>\n",
       "      <td>1711</td>\n",
       "      <td>6683</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65187</th>\n",
       "      <td>0.0008</td>\n",
       "      <td>0.0019</td>\n",
       "      <td>0.00267</td>\n",
       "      <td>2</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.046</td>\n",
       "      <td>413</td>\n",
       "      <td>9253</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>65188 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       AF_TGP  AF_ESP  AF_EXAC  IMPACT  STRAND  CADD_PHRED  SYMBOL  CLNDN\n",
       "0      0.1066  0.0771  0.10020       2     1.0       1.053     200   8960\n",
       "1      0.0000  0.0000  0.00000       2    -1.0      31.000    2135   8388\n",
       "2      0.0000  0.0000  0.00001       2    -1.0      28.100     892   8449\n",
       "3      0.0000  0.0000  0.00000       2     1.0      22.500    1889   8156\n",
       "4      0.0000  0.0000  0.00000       2     1.0      24.700    1889   8154\n",
       "...       ...     ...      ...     ...     ...         ...     ...    ...\n",
       "65183  0.1605  0.0801  0.13923       1    -1.0       0.105     707   3915\n",
       "65184  0.0013  0.0020  0.00060       2    -1.0       0.002     707   9085\n",
       "65185  0.0111  0.0125  0.00370       1    -1.0      12.850     707   9085\n",
       "65186  0.0000  0.0003  0.00034       1    -1.0       0.130    1711   6683\n",
       "65187  0.0008  0.0019  0.00267       2    -1.0       0.046     413   9253\n",
       "\n",
       "[65188 rows x 8 columns]"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# as Naive Bayes is giving very low accuracy for min max scaled data\n",
    "X =original_df[featureScores.sort_values(by='Score', ascending=False).iloc[:8,0]]\n",
    "y = original_df['CLASS']\n",
    "X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Test Accuracy 72.89461573861021\n",
      "Test Accuracy 72.89461573861021\n",
      "Test Accuracy 72.97131461880657\n",
      "Test Accuracy 73.07869305108146\n",
      "Test Accuracy 73.21675103543488\n",
      "Test Accuracy 72.83325663445314\n",
      "Test Accuracy 72.74121797821752\n",
      "Test Accuracy 73.37014879582759\n",
      "Test Accuracy 73.27401043264805\n",
      "Test Accuracy 73.64222154034981\n",
      "accuracies in iterations :  [72.89461573861021, 72.89461573861021, 72.97131461880657, 73.07869305108146, 73.21675103543488, 72.83325663445314, 72.74121797821752, 73.37014879582759, 73.27401043264805, 73.64222154034981]\n",
      "AVG:  73.09168455640395\n"
     ]
    }
   ],
   "source": [
    "# using Naive Bayes Classifier\n",
    "\n",
    "gnb = GaussianNB()\n",
    "for train_index, test_index in skf.split(X,y):\n",
    "    X_train = X.iloc[train_index]\n",
    "    y_train = y[train_index]\n",
    "    X_test = X.iloc[test_index]\n",
    "    y_test = y[test_index]\n",
    "    gnb.fit(X_train, y_train)\n",
    "    Y_testPred = gnb.predict(X_test)\n",
    "    testAccuracy = accuracy_score(y_test, Y_testPred)\n",
    "    print(\"Test Accuracy\", testAccuracy*100)\n",
    "    accsNaiveBayesClassifier.append(testAccuracy*100)\n",
    "print(\"accuracies in iterations : \",accsNaiveBayesClassifier)\n",
    "print(\"AVG: \",sum(accsNaiveBayesClassifier)/len(accsNaiveBayesClassifier))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# K NEAREST NEIGHBOUR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Test Accuracy 74.96548550391165\n",
      "Test Accuracy 74.68936953520479\n",
      "Test Accuracy 74.62801043104771\n",
      "Test Accuracy 74.87344684767604\n",
      "Test Accuracy 74.75072863936187\n",
      "Test Accuracy 74.75072863936187\n",
      "Test Accuracy 74.7814081914404\n",
      "Test Accuracy 74.87344684767604\n",
      "Test Accuracy 75.00767106474379\n",
      "Test Accuracy 74.85424976986805\n",
      "[74.96548550391165, 74.68936953520479, 74.62801043104771, 74.87344684767604, 74.75072863936187, 74.75072863936187, 74.7814081914404, 74.87344684767604, 75.00767106474379, 74.85424976986805]\n",
      "AVG:  74.81745454702921\n"
     ]
    }
   ],
   "source": [
    "from sklearn.neighbors import KNeighborsClassifier\n",
    "import math\n",
    "accsKneighbour=[]\n",
    "for train_index, test_index in skf.split(X,y):\n",
    "    X_train = X.iloc[train_index]\n",
    "    y_train = y[train_index]\n",
    "    X_test = X.iloc[test_index]\n",
    "    y_test = y[test_index]\n",
    "    KNNModel = KNeighborsClassifier(n_neighbors=int(math.sqrt(train_index.shape[0])))\n",
    "    KNNModel.fit(X_train, y_train)\n",
    "    Y_testPred = KNNModel.predict(X_test)\n",
    "    testAccuracy = accuracy_score(y_test, Y_testPred)\n",
    "    print(\"Test Accuracy\", testAccuracy*100)\n",
    "    accsKneighbour.append(testAccuracy*100)\n",
    "print(accsKneighbour)\n",
    "print(\"AVG: \",sum(accsKneighbour)/len(accsKneighbour))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
