{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# importing necessary libraries."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import warnings\n",
    "import seaborn as sns\n",
    "from sklearn.feature_selection import SelectKBest\n",
    "from sklearn.feature_selection import chi2\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.model_selection import StratifiedKFold\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.metrics import accuracy_score\n",
    "from sklearn import tree\n",
    "from sklearn.tree import DecisionTreeClassifier\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# obtaining the data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AF_ESP</th>\n",
       "      <th>AF_EXAC</th>\n",
       "      <th>AF_TGP</th>\n",
       "      <th>CLNDISDB</th>\n",
       "      <th>CLNDISDBINCL</th>\n",
       "      <th>CLNDN</th>\n",
       "      <th>...</th>\n",
       "      <th>SIFT</th>\n",
       "      <th>PolyPhen</th>\n",
       "      <th>MOTIF_NAME</th>\n",
       "      <th>MOTIF_POS</th>\n",
       "      <th>HIGH_INF_POS</th>\n",
       "      <th>MOTIF_SCORE_CHANGE</th>\n",
       "      <th>LoFtool</th>\n",
       "      <th>CADD_PHRED</th>\n",
       "      <th>CADD_RAW</th>\n",
       "      <th>BLOSUM62</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1168180</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0771</td>\n",
       "      <td>0.10020</td>\n",
       "      <td>0.1066</td>\n",
       "      <td>MedGen:CN169374</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not_specified</td>\n",
       "      <td>...</td>\n",
       "      <td>tolerated</td>\n",
       "      <td>benign</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.053</td>\n",
       "      <td>-0.208682</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1470752</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>MedGen:C1843891,OMIM:607454,Orphanet:ORPHA9877...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Spinocerebellar_ataxia_21|not_provided</td>\n",
       "      <td>...</td>\n",
       "      <td>deleterious_low_confidence</td>\n",
       "      <td>benign</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>31.000</td>\n",
       "      <td>6.517838</td>\n",
       "      <td>-3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1737942</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00001</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>Human_Phenotype_Ontology:HP:0000486,MedGen:C00...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Strabismus|Nystagmus|Hypothyroidism|Intellectu...</td>\n",
       "      <td>...</td>\n",
       "      <td>deleterious</td>\n",
       "      <td>probably_damaging</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>28.100</td>\n",
       "      <td>6.061752</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>2160305</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>MedGen:C1321551,OMIM:182212,SNOMED_CT:83092002...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Shprintzen-Goldberg_syndrome|not_provided</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22.500</td>\n",
       "      <td>3.114491</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>2160305</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>MedGen:C1321551,OMIM:182212,SNOMED_CT:83092002</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Shprintzen-Goldberg_syndrome</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>24.700</td>\n",
       "      <td>4.766224</td>\n",
       "      <td>-3.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 46 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  CHROM      POS REF ALT  AF_ESP  AF_EXAC  AF_TGP  \\\n",
       "0     1  1168180   G   C  0.0771  0.10020  0.1066   \n",
       "1     1  1470752   G   A  0.0000  0.00000  0.0000   \n",
       "2     1  1737942   A   G  0.0000  0.00001  0.0000   \n",
       "3     1  2160305   G   A  0.0000  0.00000  0.0000   \n",
       "4     1  2160305   G   T  0.0000  0.00000  0.0000   \n",
       "\n",
       "                                            CLNDISDB CLNDISDBINCL  \\\n",
       "0                                    MedGen:CN169374          NaN   \n",
       "1  MedGen:C1843891,OMIM:607454,Orphanet:ORPHA9877...          NaN   \n",
       "2  Human_Phenotype_Ontology:HP:0000486,MedGen:C00...          NaN   \n",
       "3  MedGen:C1321551,OMIM:182212,SNOMED_CT:83092002...          NaN   \n",
       "4     MedGen:C1321551,OMIM:182212,SNOMED_CT:83092002          NaN   \n",
       "\n",
       "                                               CLNDN  ...  \\\n",
       "0                                      not_specified  ...   \n",
       "1             Spinocerebellar_ataxia_21|not_provided  ...   \n",
       "2  Strabismus|Nystagmus|Hypothyroidism|Intellectu...  ...   \n",
       "3          Shprintzen-Goldberg_syndrome|not_provided  ...   \n",
       "4                       Shprintzen-Goldberg_syndrome  ...   \n",
       "\n",
       "                         SIFT           PolyPhen MOTIF_NAME MOTIF_POS  \\\n",
       "0                   tolerated             benign        NaN       NaN   \n",
       "1  deleterious_low_confidence             benign        NaN       NaN   \n",
       "2                 deleterious  probably_damaging        NaN       NaN   \n",
       "3                         NaN                NaN        NaN       NaN   \n",
       "4                         NaN                NaN        NaN       NaN   \n",
       "\n",
       "  HIGH_INF_POS MOTIF_SCORE_CHANGE  LoFtool  CADD_PHRED  CADD_RAW BLOSUM62  \n",
       "0          NaN                NaN      NaN       1.053 -0.208682      2.0  \n",
       "1          NaN                NaN      NaN      31.000  6.517838     -3.0  \n",
       "2          NaN                NaN      NaN      28.100  6.061752     -1.0  \n",
       "3          NaN                NaN      NaN      22.500  3.114491      NaN  \n",
       "4          NaN                NaN      NaN      24.700  4.766224     -3.0  \n",
       "\n",
       "[5 rows x 46 columns]"
      ]
     },
     "execution_count": 140,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "warnings.filterwarnings('ignore')\n",
    "df=pd.read_csv('clinvar_conflicting.csv')\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# data exploration."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 65188 entries, 0 to 65187\n",
      "Data columns (total 46 columns):\n",
      " #   Column              Non-Null Count  Dtype  \n",
      "---  ------              --------------  -----  \n",
      " 0   CHROM               65188 non-null  object \n",
      " 1   POS                 65188 non-null  int64  \n",
      " 2   REF                 65188 non-null  object \n",
      " 3   ALT                 65188 non-null  object \n",
      " 4   AF_ESP              65188 non-null  float64\n",
      " 5   AF_EXAC             65188 non-null  float64\n",
      " 6   AF_TGP              65188 non-null  float64\n",
      " 7   CLNDISDB            65188 non-null  object \n",
      " 8   CLNDISDBINCL        167 non-null    object \n",
      " 9   CLNDN               65188 non-null  object \n",
      " 10  CLNDNINCL           167 non-null    object \n",
      " 11  CLNHGVS             65188 non-null  object \n",
      " 12  CLNSIGINCL          167 non-null    object \n",
      " 13  CLNVC               65188 non-null  object \n",
      " 14  CLNVI               27659 non-null  object \n",
      " 15  MC                  64342 non-null  object \n",
      " 16  ORIGIN              65188 non-null  int64  \n",
      " 17  SSR                 130 non-null    float64\n",
      " 18  CLASS               65188 non-null  int64  \n",
      " 19  Allele              65188 non-null  object \n",
      " 20  Consequence         65188 non-null  object \n",
      " 21  IMPACT              65188 non-null  object \n",
      " 22  SYMBOL              65172 non-null  object \n",
      " 23  Feature_type        65174 non-null  object \n",
      " 24  Feature             65174 non-null  object \n",
      " 25  BIOTYPE             65172 non-null  object \n",
      " 26  EXON                56295 non-null  object \n",
      " 27  INTRON              8803 non-null   object \n",
      " 28  cDNA_position       56304 non-null  object \n",
      " 29  CDS_position        55233 non-null  object \n",
      " 30  Protein_position    55233 non-null  object \n",
      " 31  Amino_acids         55184 non-null  object \n",
      " 32  Codons              55184 non-null  object \n",
      " 33  DISTANCE            108 non-null    float64\n",
      " 34  STRAND              65174 non-null  float64\n",
      " 35  BAM_EDIT            31969 non-null  object \n",
      " 36  SIFT                24836 non-null  object \n",
      " 37  PolyPhen            24796 non-null  object \n",
      " 38  MOTIF_NAME          2 non-null      object \n",
      " 39  MOTIF_POS           2 non-null      float64\n",
      " 40  HIGH_INF_POS        2 non-null      object \n",
      " 41  MOTIF_SCORE_CHANGE  2 non-null      float64\n",
      " 42  LoFtool             60975 non-null  float64\n",
      " 43  CADD_PHRED          64096 non-null  float64\n",
      " 44  CADD_RAW            64096 non-null  float64\n",
      " 45  BLOSUM62            25593 non-null  float64\n",
      "dtypes: float64(12), int64(3), object(31)\n",
      "memory usage: 22.9+ MB\n"
     ]
    }
   ],
   "source": [
    "df.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>POS</th>\n",
       "      <th>AF_ESP</th>\n",
       "      <th>AF_EXAC</th>\n",
       "      <th>AF_TGP</th>\n",
       "      <th>ORIGIN</th>\n",
       "      <th>SSR</th>\n",
       "      <th>CLASS</th>\n",
       "      <th>DISTANCE</th>\n",
       "      <th>STRAND</th>\n",
       "      <th>MOTIF_POS</th>\n",
       "      <th>MOTIF_SCORE_CHANGE</th>\n",
       "      <th>LoFtool</th>\n",
       "      <th>CADD_PHRED</th>\n",
       "      <th>CADD_RAW</th>\n",
       "      <th>BLOSUM62</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>POS</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.007003</td>\n",
       "      <td>0.007721</td>\n",
       "      <td>0.004225</td>\n",
       "      <td>0.013645</td>\n",
       "      <td>0.192398</td>\n",
       "      <td>0.012061</td>\n",
       "      <td>0.197942</td>\n",
       "      <td>-0.122896</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.257226</td>\n",
       "      <td>-0.005409</td>\n",
       "      <td>-0.009753</td>\n",
       "      <td>0.022451</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AF_ESP</th>\n",
       "      <td>0.007003</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.851870</td>\n",
       "      <td>0.807741</td>\n",
       "      <td>-0.013563</td>\n",
       "      <td>-0.077253</td>\n",
       "      <td>-0.126538</td>\n",
       "      <td>0.295624</td>\n",
       "      <td>-0.011551</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.027343</td>\n",
       "      <td>-0.164169</td>\n",
       "      <td>-0.148118</td>\n",
       "      <td>0.039367</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AF_EXAC</th>\n",
       "      <td>0.007721</td>\n",
       "      <td>0.851870</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.805634</td>\n",
       "      <td>-0.013432</td>\n",
       "      <td>0.048413</td>\n",
       "      <td>-0.121213</td>\n",
       "      <td>0.291840</td>\n",
       "      <td>-0.013358</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.027831</td>\n",
       "      <td>-0.154891</td>\n",
       "      <td>-0.140803</td>\n",
       "      <td>0.038996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AF_TGP</th>\n",
       "      <td>0.004225</td>\n",
       "      <td>0.807741</td>\n",
       "      <td>0.805634</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.014450</td>\n",
       "      <td>-0.114526</td>\n",
       "      <td>-0.130077</td>\n",
       "      <td>0.040676</td>\n",
       "      <td>-0.014002</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.029447</td>\n",
       "      <td>-0.167057</td>\n",
       "      <td>-0.150989</td>\n",
       "      <td>0.043167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORIGIN</th>\n",
       "      <td>0.013645</td>\n",
       "      <td>-0.013563</td>\n",
       "      <td>-0.013432</td>\n",
       "      <td>-0.014450</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.175675</td>\n",
       "      <td>0.004357</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.007564</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.028363</td>\n",
       "      <td>0.060290</td>\n",
       "      <td>0.064609</td>\n",
       "      <td>-0.012681</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              POS    AF_ESP   AF_EXAC    AF_TGP    ORIGIN       SSR     CLASS  \\\n",
       "POS      1.000000  0.007003  0.007721  0.004225  0.013645  0.192398  0.012061   \n",
       "AF_ESP   0.007003  1.000000  0.851870  0.807741 -0.013563 -0.077253 -0.126538   \n",
       "AF_EXAC  0.007721  0.851870  1.000000  0.805634 -0.013432  0.048413 -0.121213   \n",
       "AF_TGP   0.004225  0.807741  0.805634  1.000000 -0.014450 -0.114526 -0.130077   \n",
       "ORIGIN   0.013645 -0.013563 -0.013432 -0.014450  1.000000  0.175675  0.004357   \n",
       "\n",
       "         DISTANCE    STRAND  MOTIF_POS  MOTIF_SCORE_CHANGE   LoFtool  \\\n",
       "POS      0.197942 -0.122896        NaN                -1.0  0.257226   \n",
       "AF_ESP   0.295624 -0.011551        NaN                 NaN  0.027343   \n",
       "AF_EXAC  0.291840 -0.013358        NaN                 NaN  0.027831   \n",
       "AF_TGP   0.040676 -0.014002        NaN                 NaN  0.029447   \n",
       "ORIGIN        NaN -0.007564        NaN                 NaN -0.028363   \n",
       "\n",
       "         CADD_PHRED  CADD_RAW  BLOSUM62  \n",
       "POS       -0.005409 -0.009753  0.022451  \n",
       "AF_ESP    -0.164169 -0.148118  0.039367  \n",
       "AF_EXAC   -0.154891 -0.140803  0.038996  \n",
       "AF_TGP    -0.167057 -0.150989  0.043167  \n",
       "ORIGIN     0.060290  0.064609 -0.012681  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(df.corr(numeric_only=True).head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# dropping the useless columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.drop(['CLNDISDBINCL','CLNDNINCL','CLNSIGINCL','CLNVI','SSR','INTRON','DISTANCE','BAM_EDIT','SIFT','PolyPhen','MOTIF_NAME','MOTIF_POS','HIGH_INF_POS','MOTIF_SCORE_CHANGE','BLOSUM62'],axis=1,inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CHROM                   0\n",
       "POS                     0\n",
       "REF                     0\n",
       "ALT                     0\n",
       "AF_ESP                  0\n",
       "AF_EXAC                 0\n",
       "AF_TGP                  0\n",
       "CLNDISDB                0\n",
       "CLNDN                   0\n",
       "CLNHGVS                 0\n",
       "CLNVC                   0\n",
       "MC                    846\n",
       "ORIGIN                  0\n",
       "CLASS                   0\n",
       "Allele                  0\n",
       "Consequence             0\n",
       "IMPACT                  0\n",
       "SYMBOL                 16\n",
       "Feature_type           14\n",
       "Feature                14\n",
       "BIOTYPE                16\n",
       "EXON                 8893\n",
       "cDNA_position        8884\n",
       "CDS_position         9955\n",
       "Protein_position     9955\n",
       "Amino_acids         10004\n",
       "Codons              10004\n",
       "STRAND                 14\n",
       "LoFtool              4213\n",
       "CADD_PHRED           1092\n",
       "CADD_RAW             1092\n",
       "dtype: int64"
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.isnull().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0   -1.0\n",
       "Name: STRAND, dtype: float64"
      ]
     },
     "execution_count": 145,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['STRAND'].mode()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['LoFtool'].fillna(int(df['LoFtool'].mean()), inplace=True)\n",
    "df['CADD_PHRED'].fillna(int(df['CADD_PHRED'].mean()), inplace=True)\n",
    "df['CADD_RAW'].fillna(int(df['CADD_RAW'].mean()), inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['STRAND'].fillna(df['STRAND'].mode()[0],inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CHROM                   0\n",
       "POS                     0\n",
       "REF                     0\n",
       "ALT                     0\n",
       "AF_ESP                  0\n",
       "AF_EXAC                 0\n",
       "AF_TGP                  0\n",
       "CLNDISDB                0\n",
       "CLNDN                   0\n",
       "CLNHGVS                 0\n",
       "CLNVC                   0\n",
       "MC                    846\n",
       "ORIGIN                  0\n",
       "CLASS                   0\n",
       "Allele                  0\n",
       "Consequence             0\n",
       "IMPACT                  0\n",
       "SYMBOL                 16\n",
       "Feature_type           14\n",
       "Feature                14\n",
       "BIOTYPE                16\n",
       "EXON                 8893\n",
       "cDNA_position        8884\n",
       "CDS_position         9955\n",
       "Protein_position     9955\n",
       "Amino_acids         10004\n",
       "Codons              10004\n",
       "STRAND                  0\n",
       "LoFtool                 0\n",
       "CADD_PHRED              0\n",
       "CADD_RAW                0\n",
       "dtype: int64"
      ]
     },
     "execution_count": 148,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.isnull().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [],
   "source": [
    "object_columns = df.select_dtypes(['object']).columns\n",
    "object_columns\n",
    "for c in df.columns:\n",
    "    if c in object_columns:\n",
    "        df[[c]]=df[[c]].astype('category')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['CHROM', 'REF', 'ALT', 'CLNDISDB', 'CLNDN', 'CLNHGVS', 'CLNVC', 'MC',\n",
      "       'Allele', 'Consequence', 'IMPACT', 'SYMBOL', 'Feature_type', 'Feature',\n",
      "       'BIOTYPE', 'EXON', 'cDNA_position', 'CDS_position', 'Protein_position',\n",
      "       'Amino_acids', 'Codons'],\n",
      "      dtype='object')\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Index(['CHROM', 'REF', 'ALT', 'CLNDISDB', 'CLNDN', 'CLNHGVS', 'CLNVC', 'MC',\n",
       "       'Allele', 'Consequence', 'IMPACT', 'SYMBOL', 'Feature_type', 'Feature',\n",
       "       'BIOTYPE', 'EXON', 'cDNA_position', 'CDS_position', 'Protein_position',\n",
       "       'Amino_acids', 'Codons'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 150,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cat_columns = df.select_dtypes(['category']).columns\n",
    "print(cat_columns)\n",
    "df[cat_columns] = df[cat_columns].apply(lambda x: x.cat.codes)\n",
    "cat_columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CHROM                  int8\n",
       "POS                   int64\n",
       "REF                   int16\n",
       "ALT                   int16\n",
       "AF_ESP              float64\n",
       "AF_EXAC             float64\n",
       "AF_TGP              float64\n",
       "CLNDISDB              int16\n",
       "CLNDN                 int16\n",
       "CLNHGVS               int32\n",
       "CLNVC                  int8\n",
       "MC                     int8\n",
       "ORIGIN                int64\n",
       "CLASS                 int64\n",
       "Allele                int16\n",
       "Consequence            int8\n",
       "IMPACT                 int8\n",
       "SYMBOL                int16\n",
       "Feature_type           int8\n",
       "Feature               int16\n",
       "BIOTYPE                int8\n",
       "EXON                  int16\n",
       "cDNA_position         int16\n",
       "CDS_position          int16\n",
       "Protein_position      int16\n",
       "Amino_acids           int16\n",
       "Codons                int16\n",
       "STRAND              float64\n",
       "LoFtool             float64\n",
       "CADD_PHRED          float64\n",
       "CADD_RAW            float64\n",
       "dtype: object"
      ]
     },
     "execution_count": 151,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CHROM               0.002030\n",
       "POS                 0.012061\n",
       "REF                 0.000586\n",
       "ALT                -0.000748\n",
       "AF_ESP             -0.126538\n",
       "AF_EXAC            -0.121213\n",
       "AF_TGP             -0.130077\n",
       "CLNDISDB            0.012407\n",
       "CLNDN               0.023176\n",
       "CLNHGVS            -0.013249\n",
       "CLNVC               0.037792\n",
       "MC                  0.005494\n",
       "ORIGIN              0.004357\n",
       "CLASS               1.000000\n",
       "Allele              0.007442\n",
       "Consequence        -0.001063\n",
       "IMPACT              0.068912\n",
       "SYMBOL              0.021568\n",
       "Feature_type       -0.014732\n",
       "Feature             0.001521\n",
       "BIOTYPE            -0.013692\n",
       "EXON               -0.012167\n",
       "cDNA_position      -0.011147\n",
       "CDS_position       -0.013195\n",
       "Protein_position   -0.007403\n",
       "Amino_acids        -0.017456\n",
       "Codons             -0.018030\n",
       "STRAND             -0.031551\n",
       "LoFtool            -0.002226\n",
       "CADD_PHRED         -0.037178\n",
       "CADD_RAW           -0.052208\n",
       "Name: CLASS, dtype: float64"
      ]
     },
     "execution_count": 152,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.corr()['CLASS']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [],
   "source": [
    "a=['CLNDN','CLNVC','IMPACT','SYMBOL']#list of columns having negative values but may be relevant , interpreted from correlation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "16"
      ]
     },
     "execution_count": 154,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(df['SYMBOL'] < 0).sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [],
   "source": [
    "# selecting only those columns which are positive As chi square cant handle negative values\n",
    "li=df.columns[(df>=0).all()].tolist()\n",
    "li\n",
    "df_positive=df[li]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# feature selection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "         Specs         Score\n",
      "1          POS  4.362602e+08\n",
      "9      CLNHGVS  1.243176e+05\n",
      "8        CLNDN  5.431848e+04\n",
      "7     CLNDISDB  1.306732e+04\n",
      "16  CADD_PHRED  6.637613e+02\n",
      "12      Allele  3.907656e+02\n",
      "6       AF_TGP  2.560591e+02\n",
      "4       AF_ESP  2.402746e+02\n"
     ]
    }
   ],
   "source": [
    "# using chi square for feature selection\n",
    "data=df_positive.copy()\n",
    "\n",
    "X = data.loc[:,data.columns!='CLASS']\n",
    "y = data.loc[:,'CLASS']\n",
    "bestfeatures = SelectKBest(score_func=chi2, k='all')\n",
    "fit = bestfeatures.fit(X,y)\n",
    "dfscores = pd.DataFrame(fit.scores_)\n",
    "dfcolumns = pd.DataFrame(X.columns)\n",
    "featureScores = pd.concat([dfcolumns,dfscores],axis=1)\n",
    "featureScores.columns = ['Specs','Score']\n",
    "print(featureScores.nlargest(8,'Score'))#printing the 8 best features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1168180</td>\n",
       "      <td>326</td>\n",
       "      <td>8960</td>\n",
       "      <td>8866</td>\n",
       "      <td>1.053</td>\n",
       "      <td>8960</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1470752</td>\n",
       "      <td>467</td>\n",
       "      <td>8388</td>\n",
       "      <td>6615</td>\n",
       "      <td>31.000</td>\n",
       "      <td>8388</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>2135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1737942</td>\n",
       "      <td>1179</td>\n",
       "      <td>8449</td>\n",
       "      <td>750</td>\n",
       "      <td>28.100</td>\n",
       "      <td>8449</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2160305</td>\n",
       "      <td>1826</td>\n",
       "      <td>8156</td>\n",
       "      <td>5993</td>\n",
       "      <td>22.500</td>\n",
       "      <td>8156</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>1889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2160305</td>\n",
       "      <td>1827</td>\n",
       "      <td>8154</td>\n",
       "      <td>5989</td>\n",
       "      <td>24.700</td>\n",
       "      <td>8154</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>1889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65183</th>\n",
       "      <td>154158201</td>\n",
       "      <td>64090</td>\n",
       "      <td>3915</td>\n",
       "      <td>3195</td>\n",
       "      <td>0.105</td>\n",
       "      <td>3915</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>707</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65184</th>\n",
       "      <td>154159118</td>\n",
       "      <td>64093</td>\n",
       "      <td>9085</td>\n",
       "      <td>8932</td>\n",
       "      <td>0.002</td>\n",
       "      <td>9085</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>707</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65185</th>\n",
       "      <td>154194886</td>\n",
       "      <td>64096</td>\n",
       "      <td>9085</td>\n",
       "      <td>8932</td>\n",
       "      <td>12.850</td>\n",
       "      <td>9085</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>707</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65186</th>\n",
       "      <td>154490187</td>\n",
       "      <td>64097</td>\n",
       "      <td>6683</td>\n",
       "      <td>8221</td>\n",
       "      <td>0.130</td>\n",
       "      <td>6683</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>1711</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65187</th>\n",
       "      <td>154508542</td>\n",
       "      <td>64098</td>\n",
       "      <td>9253</td>\n",
       "      <td>9156</td>\n",
       "      <td>0.046</td>\n",
       "      <td>9253</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>413</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>65188 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               0      1     2     3       4     5  6  7     8\n",
       "0        1168180    326  8960  8866   1.053  8960  6  2   200\n",
       "1        1470752    467  8388  6615  31.000  8388  6  2  2135\n",
       "2        1737942   1179  8449   750  28.100  8449  6  2   892\n",
       "3        2160305   1826  8156  5993  22.500  8156  6  2  1889\n",
       "4        2160305   1827  8154  5989  24.700  8154  6  2  1889\n",
       "...          ...    ...   ...   ...     ...   ... .. ..   ...\n",
       "65183  154158201  64090  3915  3195   0.105  3915  6  1   707\n",
       "65184  154159118  64093  9085  8932   0.002  9085  6  2   707\n",
       "65185  154194886  64096  9085  8932  12.850  9085  6  1   707\n",
       "65186  154490187  64097  6683  8221   0.130  6683  6  1  1711\n",
       "65187  154508542  64098  9253  9156   0.046  9253  6  2   413\n",
       "\n",
       "[65188 rows x 9 columns]"
      ]
     },
     "execution_count": 157,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X = data[featureScores.sort_values(by='Score', ascending=False).iloc[:5,0]]\n",
    "X=pd.concat([X,df[a]],ignore_index=True,sort=False,axis=1)\n",
    "y = data['CLASS']\n",
    "X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "metadata": {},
   "outputs": [],
   "source": [
    "# scaling the data\n",
    "scaler = StandardScaler()\n",
    "XScaled = scaler.fit_transform(X)\n",
    "accs = []\n",
    "skf = StratifiedKFold(n_splits=10,random_state=1, shuffle=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {},
   "outputs": [],
   "source": [
    "def logisticRegression(X,y):\n",
    "    for train_index, test_index in skf.split(X,y):\n",
    "        X_train = X.iloc[train_index]\n",
    "        y_train = y[train_index]\n",
    "        X_test = X.iloc[test_index]\n",
    "        y_test = y[test_index]\n",
    "        LRModel = LogisticRegression(penalty = 'l2',max_iter=500)\n",
    "        LRModel.fit(X_train, y_train)\n",
    "        Y_testPred = LRModel.predict(X_test)\n",
    "        testAccuracy = accuracy_score(y_test, Y_testPred)\n",
    "        accs.append(testAccuracy*100)\n",
    "    print(\"Test Accuracy\", testAccuracy*100)\n",
    "    print(accs)\n",
    "    print(\"AVG: \",sum(accs)/len(accs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Test Accuracy 74.79674796747967\n",
      "Test Accuracy 74.79674796747967\n",
      "Test Accuracy 74.79674796747967\n",
      "Test Accuracy 74.79674796747967\n",
      "Test Accuracy 74.7814081914404\n",
      "Test Accuracy 74.7814081914404\n",
      "Test Accuracy 74.7814081914404\n",
      "Test Accuracy 74.7814081914404\n",
      "Test Accuracy 74.79288125191776\n",
      "Test Accuracy 74.79288125191776\n",
      "[74.79674796747967, 74.79674796747967, 74.79674796747967, 74.79674796747967, 74.7814081914404, 74.7814081914404, 74.7814081914404, 74.7814081914404, 74.79288125191776, 74.79288125191776]\n",
      "AVG:  74.78983871395158\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# using logistic regression model.\n",
    "logisticRegression(X,y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
